opal/mca/base/mca_base_source.c

```diff
@@ -0,0 +1,242 @@
+    if (0 == opal_clock_getres(&spec)){
+        return (uint64_t)(spec.tv_sec + spec.tv_nsec);
+    } else {
+        /* guess */
+        return 1000000;
+    }
+}
+
+/***************************************************************************************************/
+
+int mca_base_source_init (void)
+{
+    int ret = OPAL_SUCCESS;
+
+    OPAL_THREAD_LOCK(&mca_base_source_lock);
+    if (false == mca_base_source_initialized) {
+        mca_base_source_initialized = true;
+
+        OBJ_CONSTRUCT(&registered_sources, opal_pointer_array_t);
+        opal_pointer_array_init(&registered_sources, 16, 512, 16);
+
+        mca_base_source_default_source = mca_base_source_register ("opal", "mca", "base", "default_source",
+                                                                   "Default source for MCA events", true,
+                                                                   mca_base_source_default_time_source,
+                                                                   mca_base_source_default_time_source_ticks ());
+
+    }
+    OPAL_THREAD_UNLOCK(&mca_base_source_lock);
+
+    return ret;
+}
+
+int mca_base_source_finalize (void)
+{
+    int i;
+
+    OPAL_THREAD_LOCK(&mca_base_source_lock);
+    if (true == mca_base_source_initialized)  {
+        for (i = 0 ; i < source_count ; ++i) {
+            mca_base_source_t *source = opal_pointer_array_get_item (&registered_sources, i);
+            if (source) {
+                OBJ_RELEASE(source);
+            }
+        }
+
+        source_count = 0;
+
+        OBJ_DESTRUCT(&registered_sources);
+        mca_base_source_initialized = false;
+    }
+    OPAL_THREAD_UNLOCK(&mca_base_source_lock);
+
+    return OPAL_SUCCESS;
+}
+
+/***************************************************************************************************/
+
+mca_base_source_t *mca_base_source_get (int source_index)
+{
+    return opal_pointer_array_get_item (&registered_sources, source_index);
+}
+
+int mca_base_source_set_time_source (int source_index, mca_base_source_time_fn_t time_source, uint64_t time_ticks)
+{
+    mca_base_source_t *source = mca_base_source_get (source_index);
+
+    if (NULL == source) {
+        return OPAL_ERR_NOT_FOUND;
+    }
+
+    if (!time_source) {
+        time_source = mca_base_source_default_time_source;
+        time_ticks = mca_base_source_default_time_source_ticks ();
+    }
+
+    source->source_time = time_source;
+    source->source_ticks = time_ticks;
+
+    return OPAL_SUCCESS;
+}
+
+/***************************************************************************************************/
+
+int mca_base_source_get_count (int *count)
+{
+    *count = source_count;
+    return OPAL_SUCCESS;
+}
+
+static inline int mca_base_source_get_by_name (const char *name, mca_base_source_t **source_out)
+{
+    /* there are expected to be a relatively small number of sources so a linear search should be fine */
+    for (int i = 0 ; i < source_count ; ++i) {
```

Comment:
Shouldn't this routine be protected by the lock?  E.g., what if another thread changes `source_count`, or `registered_sources`? (or can that never happen?)

Reply:
Same here, too, source info won't change.

Reply:
Per MPI-4.0 p758 21:22, the number of sources is allowed to increase during the execution of an MPI process.

Reply:
I am adding locking to protect source access and prevent potential future source access issues.

Reply:
Addressed with https://github.com/open-mpi/ompi/commit/c9b35e2ac9a780d16a05bfc9b82b4aef5ab0a970